File: Stage29/Python/Problem13549.py
import sys
from collections import deque

input = sys.stdin.readline
MAX = 100001

def bfs(start, target):
    dist = [float('inf')] * MAX
    dq = deque()

    dq.append(start)
    dist[start] = 0

    while dq:
        cur = dq.popleft()

        if cur == target:
            return dist[cur]

        next_pos = cur * 2
        if next_pos < MAX and dist[next_pos] > dist[cur]:
            dist[next_pos] = dist[cur]
            dq.appendleft(next_pos)

        next_pos = cur - 1
        if next_pos >= 0 and dist[next_pos] > dist[cur] + 1:
            dist[next_pos] = dist[cur] + 1
            dq.append(next_pos)

        next_pos = cur + 1
        if next_pos < MAX and dist[next_pos] > dist[cur] + 1:
            dist[next_pos] = dist[cur] + 1
            dq.append(next_pos)

def main():
    N, K = map(int, input().split())
    print(bfs(N, K))

if __name__ == "__main__":
    main()

File: Stage29/Java/Problem13549.java
package Stage29.Java;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Deque;
import java.util.StringTokenizer;

public class Problem13549 {
    static final int MAX = 100001;
    static int[] dist = new int[MAX];

    static int bfs(int start, int target) {
        Deque<Integer> dq = new ArrayDeque<>();

        dq.add(start);
        dist[start] = 0;

        while (!dq.isEmpty()) {
            int cur = dq.pollFirst();

            if (cur == target) {
                return dist[cur];
            }

            int next = cur * 2;
            if (next < MAX && dist[next] > dist[cur]) {
                dist[next] = dist[cur];
                dq.addFirst(next);
            }

            next = cur - 1;
            if (next >= 0 && dist[next] > dist[cur] + 1) {
                dist[next] = dist[cur] + 1;
                dq.addLast(next);
            }

            next = cur + 1;
            if (next < MAX && dist[next] > dist[cur] + 1) {
                dist[next] = dist[cur] + 1;
                dq.addLast(next);
            }
        }
        
        return -1;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());

        Arrays.fill(dist, Integer.MAX_VALUE);

        int result = bfs(N, K);

        System.out.println(result);
        br.close();
    }
}
